mlir/lib/Dialect/Rock/Transforms/AnnotateLiveness.cpp

```
@@ -0,0 +1,305 @@
+      accesses.push_back(op);
+  });
+
+  return accesses;
+}
+
+// Compute live ranges based on write/read pattern
+// See the comment in annotateLiveness() to understand the assumptions we make
+// here.
+static FailureOr<SmallVector<LiveRange>> computeLiveRanges(GpuAllocOp buffer,
+                                                           func::FuncOp func) {
+  SmallVector<LiveRange> ranges;
+
+  // Get all accesses in program order
+  SmallVector<Operation *> accesses = getOrderedAccesses(buffer, func);
+
+  if (accesses.empty())
+    return ranges;
+
+  // State machine to track write/read patterns
+  Operation *currentWrite = nullptr;
+  Operation *lastRead = nullptr;
+
+  for (Operation *op : accesses) {
+    bool isWrite = hasWriteEffect(op, buffer);
+    bool isRead = hasReadEffect(op, buffer);
+    assert(!(isWrite && isRead) &&
+           "We do not expect to have an op that reads and writes to LDS at "
+           "this point (AnnotateLiveness)");
+
+    if (isWrite) {
+      // If we have a pending range (write followed by reads), close it
+      if (currentWrite && lastRead) {
+        LiveRange range{currentWrite, lastRead};
+        ranges.emplace_back(range);
+        currentWrite = nullptr;
+        lastRead = nullptr;
+      }
+
+      // Start a new range if we don't have one
+      if (!currentWrite) {
+        currentWrite = op;
+      }
+      // If we already have a write, this is another write in the same range
+      // (write, write, ... pattern) - keep the first write as range start
+    }
+
+    if (isRead) {
+      // Update the last read (could be write, read, read, ... pattern)
+      lastRead = op;
+      if (!currentWrite) {
+        return buffer->emitError("Read before write");
+      }
+    }
+  }
+
+  bool hasRead = lastRead != nullptr;
+  bool hasWrite = currentWrite != nullptr;
+  if (hasRead != hasWrite) {
+    return buffer->emitError("Found a non closed read-write pattern");
```

I'm not sure I understand this. Let's take `non_closed_read_write_pattern` as an example.

Yes there is a write on a buffer that is not read later, but we could have a valid program that does that, right? It would be cleaned by dce at some point.

I guess supporting that complicates the logic, but can't we treat the `return` op as an implicit `live.live_out` for all buffers?

---

> Yes there is a write on a buffer that is not read later, but we could have a valid program that does that, right? It would be cleaned by dce at some point.

it's valid but it doesn't really make sense to write to LDS if you won't read from it later. Also, I'm not sure the compiler is able to figure out that it can be removed (because it's LDS).

> I guess supporting that complicates the logic, but can't we treat the return op as an implicit live.live_out for all buffers?

We have now multiple gemm in the same kernel (for backward convs). So, if one gemm left one range open (writing to LDS and not reading) it would affect the next gemm by not letting it reuse the memory. I'd prefer to fail on those cases rather than silently close the liveness range on return. 

I think our advantage is that our code generation is simple and we can impose these kind of simple patterns and use the assumptions to our advantage. If at some point we need to violate the write() -> read() we can always change the code, right?

---

I agree, but in my opinion reading the code is not enough to conclude all that you said. You can conclude that such case is not supported, but all the second paragraph that you mentioned is missed. It might be a fair point to do this, but it's not trivial to conclude from the code.

At least for me that complicates understanding the code +500%. In a ideal world we could have a document that talks about all these "assumptions".

---

Yes, I agree, we could create a confluence page for each pass where we write its assumptions and when it would stop working etc. I can start for annotate liveness, but we should do it for every pass.

---

https://amd.atlassian.net/wiki/spaces/MLSE/pages/1172542715/Annotate+Liveness+Pass